daemon/input.c

```diff
@@ -0,0 +1,303 @@
+					index--;
+				}
+			} else if ((keyboard_event.code == KEY_ENTER) ||
+				   (keyboard_event.code == KEY_KPENTER)) {
+				// Enter was pressed, finish
+				if (!key) {
+					key = (char *)mem_alloc(sizeof(char));
+					*key = '\0';
+				}
+				index = 0;
+				DEBUG("Starting container with smartcard from callback");
+				if (cmld_container_start_with_smartcard(
+					    cb_data->control, cb_data->container, key) != 0) {
+					ERROR("Failed to start container %s",
+					      container_get_name(cb_data->container));
+				}
+				goto exit;
+			} else {
+				// Another key was pressed..
+				char c;
+				// ..Check if pressed key is a digit and store it
+				if (input_event_code_to_ascii_num(keyboard_event.code, &c) == 0) {
+					index++;
+					tmp_key = (char *)mem_realloc(key, index * sizeof(char));
+					if (tmp_key) {
+						key = tmp_key;
+						key[index - 1] = c;
+					} else {
+						WARN("Failed to allocate memory for pin");
+						free(key);
+						index = 0;
+						goto exit;
+					}
+				}
+			}
+		}
+	}
+
+	// Pin entry is not yet finished, just return to wait for more keystrokes
+	return;
+
+exit:
+	if (ioctl(fd, EVIOCGRAB, NULL) != 0) {
+		WARN("Failed to release exclusive access to pin reader");
+	}
+	event_remove_io(io);
+	event_io_free(io);
+	close(fd);
+	mem_free(data);
+	if (key) {
+		memset(key, 0x0, strlen(key));
+		mem_free(key);
+	}
+	return;
+}
+
+int
+input_register_container_start_cb(control_t *control, container_t *container,
+				  ContainerStartParams *start_params)
+{
+	char *input_file = NULL;
+	int res = -1;
+
+	TRACE("Searching for USB pin reader for interactive pin entry");
+
+	// Iterate through usb-dev list and look for USB_PIN_ENTRY device
+	uevent_usbdev_t *usbdev_pinreader = NULL;
+	for (list_t *l = container_get_usbdev_list(container); l; l = l->next) {
+		uevent_usbdev_t *usbdev = (uevent_usbdev_t *)l->data;
+		if (uevent_usbdev_get_type(usbdev) == UEVENT_USBDEV_TYPE_PIN_ENTRY) {
+			usbdev_pinreader = usbdev;
+			break;
+		}
+	}
+
+	if (!usbdev_pinreader) {
+		ERROR("Failed to find USB pin reader");
+		res = -1;
+		goto exit;
+	}
+
+	TRACE("Found USB pin reader. Device Serial: %s. Vendor:Product: %x:%x",
+	      uevent_usbdev_get_i_serial(usbdev_pinreader),
+	      uevent_usbdev_get_id_vendor(usbdev_pinreader),
+	      uevent_usbdev_get_id_product(usbdev_pinreader));
+
+	input_file = input_get_usb_input_event_file(uevent_usbdev_get_id_vendor(usbdev_pinreader),
+						    uevent_usbdev_get_id_product(usbdev_pinreader));
+	if (!input_file) {
+		ERROR("Failed to find USB pin reader input file");
+		res = -1;
+		goto exit;
+	}
+
+	TRACE("Found USB pin reader input file: %s", input_file);
+
+	int fd = open(input_file, O_RDONLY | O_NONBLOCK);
+	if (fd == -1) {
+		WARN("Failed to open usb input event file for pin reader");
+		goto exit;
+	}
+
+	int ret = ioctl(fd, EVIOCGRAB, (void *)1);
+	if (ret != 0) {
+		WARN("Failed to get exclusive access to pin reader");
+		close(fd);
+		goto exit;
+	}
+
```

Comment:
make fd nonblocking for async read before registering in event loop
"fd_make_non_blocking(fd)" see common/fd.h

Reply:
As discussed: Already in open call